### PR TITLE
PR

### PR DESCRIPTION
**Pull request by D3N**

### CODE REVIEW

File: lib/factory.rb [hunk 1]
```diff
@@ -14,3 +16,100 @@
...
+
+class Factory
+  class << self
+    def new(*args, &block)
+      if args.first.instance_of? String
+        name = args.first
+        args.shift
+        const_set(name, klass(*args, &block))
+      else
+        klass(*args, &block)
```
Comment (yuriy1991) [inline lib/factory.rb]: I think `create_klass` is more obvious name

File: lib/factory.rb [hunk 2]
```diff
@@ -14,3 +16,100 @@
...
+        class_eval(&block) if block_given?
+
+        define_method :initialize do |*instance_args|
+          raise ArgumentError, 'Check arguments quantity' unless args.size == instance_args.size
+
+          args.zip(instance_args).to_h.each { |key, val| instance_variable_set("@#{key}", val) }
+        end
+
+        define_method :== do |other|
+          self_values = instance_variables.map { |var| instance_variable_get(var) }
```
Comment (yuriy1991) [inline lib/factory.rb]: can we use `values` method that you define below? it does the same thing

File: lib/factory.rb [hunk 3]
```diff
@@ -14,3 +16,100 @@
...
+          other_values = other.instance_variables.map { |var| other.instance_variable_get(var) }
+          return true if self_values == other_values
+
+          false
+        end
+
+        define_method :[] do |arg|
+          case arg
+          when String  then instance_variable_get("@#{arg.to_sym}")
```
Comment (yuriy1991) [inline lib/factory.rb]: I think you can use `@#{arg}` for strings (without to_sym)

File: lib/factory.rb [hunk 4]
```diff
@@ -14,3 +16,100 @@
...
+          when Symbol  then instance_variable_get("@#{arg}")
+          when Integer then instance_variable_get(instance_variables[arg])
+          end
+        end
+
+        define_method :[]= do |arg, new_value|
+          case arg
+          when String  then instance_variable_set("@#{arg.to_sym}", new_value)
```
Comment (yuriy1991) [inline lib/factory.rb]: I think you can use @#{arg} for strings (without to_sym)

File: lib/factory.rb [hunk 5]
```diff
@@ -14,3 +16,100 @@
...
+          when Integer then instance_variable_set(instance_variables[arg], new_value)
+          end
+        end
+
+        define_method :values do
+          instance_variables.map { |arg| instance_variable_get(arg) }
+        end
+
+        define_method :values_at do |*indexes|
+          indexes.map { |ind| values[ind] }
```
Comment (yuriy1991) [inline lib/factory.rb]: Better not to use abbreviations. Please replace `ids` with `index`

File: lib/factory.rb [hunk 6]
```diff
@@ -14,3 +16,100 @@
...
+
+        define_method :members do
+          instance_variables.map { |var| var.to_s.gsub(/@/, '').to_sym }
+        end
+
+        define_method :length do
+          values.count
+        end
+
+        define_method :each do |&blok|
```
[the remark this answers is not in the document]
Comment (D3N) [reply lib/factory.rb]: @yuriy1991 thx, fixed all.

File: lib/factory.rb [hunk 7]
```diff
@@ -20,62 +20,56 @@
...
 class Factory
   class << self
-    def new(*args, &block)
-      if args.first.instance_of? String
-        name = args.first
-        args.shift
-        const_set(name, klass(*args, &block))
+    def new(*arguments, &block)
+      if arguments.first.instance_of? String
+        name = arguments.first
```
Comment (yuriy1991) [inline lib/factory.rb]: you can do the arguments splitting in one line: `name, *rest = arguments`

Comment (D3N) [reply lib/factory.rb]: @yuriy1991 didn't know such approach, thanks, very useful.

**Pull request merged**